src/main/java/com/hivemq/persistence/local/memory/ClientSessionSubscriptionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,216 @@
+
+        buckets[bucketIndex].computeIfPresent(client, (ignore, entry) -> {
+            final ImmutableSet.Builder<Topic> remainingTopicsBuilder = new ImmutableSet.Builder<>();
+            boolean remaining = false;
+            for (final Topic topic : entry.getObject()) {
+                if (topics.contains(topic.getTopic())) {
+                    currentMemorySize.addAndGet(-topic.getEstimatedSize());
+                    continue;
+                }
+                remaining = true;
+                remainingTopicsBuilder.add(topic);
+            }
+            if (!remaining) {
+                currentMemorySize.addAndGet(-IterablePersistenceEntry.getFixedSize());
+                currentMemorySize.addAndGet(-ObjectMemoryEstimation.stringSize(client));
+                return null;
+            }
+            return new IterablePersistenceEntry<>(remainingTopicsBuilder.build(), timestamp);
+        });
+    }
+
+    @Override
+    @ExecuteInSingleWriter
+    public void removeAll(@NotNull final String client, final long timestamp, final int bucketIndex) {
+        checkNotNull(client, "Clientid must not be null");
+        checkState(timestamp > 0, "Timestamp must not be 0");
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+        final IterablePersistenceEntry<ImmutableSet<Topic>> remove = buckets[bucketIndex].remove(client);
+        if(remove == null){
+            return;
+        }
+        currentMemorySize.addAndGet(-remove.getEstimatedSize());
+        currentMemorySize.addAndGet(-ObjectMemoryEstimation.stringSize(client));
+    }
+
+    @Override
+    @NotNull
+    public ImmutableSet<Topic> getSubscriptions(@NotNull final String client) {
+        checkNotNull(client, "Clientid must not be null");
+
+        final IterablePersistenceEntry<ImmutableSet<Topic>> entry =
+                buckets[BucketUtils.getBucket(client, bucketCount)].get(client);
+        if(entry == null){
+            return ImmutableSet.of();
+        } else {
+            return entry.getObject();
+        }
+
+    }
+
+    @Override
+    @NotNull
+    public BucketChunkResult<Map<String, ImmutableSet<Topic>>> getAllSubscribersChunk(
+            final int bucketIndex,
+            @Nullable final String lastClientIdIgnored,
+            final int maxResultsIgnored) {
+
+        //as all subscriptions are already in memory, we can ignore any pagination here and return the whole bucket.
+        final Map<String, ImmutableSet<Topic>> result = buckets[bucketIndex].entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getObject()));
+
+        return new BucketChunkResult<>(result, true, lastClientIdIgnored, bucketIndex);
+
+    }
+
+    @Override
+    public void cleanUp(final int bucket) {
+        //noop because we have no duplicates in memory
+    }
+
+    @Override
+    public void closeDB(final int bucketIndex) {
+        //noop because closing is JVMs job
```

Comment:
As we have an embedded mode now, the comment's statement is not true anymore. This needs a cleanup imho.
clearing the bucket's map might be enough

Reply:
nice catch